### DISCUSSION TITLE
Enhancement: 3DS Song of Double Time Behavior

### DISCUSSION THREAD

**Discussion by Dumbledork01** [HarbourMasters/2ship2harkinian#600] [Ideas]
Hey ya'll, great work on this decomp project!

I wanted to put this feature request out there since this is one of the only features from the 3DS port that was actually good.
I know ya'll have the developer tools already that allows selecting the current time (somewhere in SaveEditor.cpp), but I think an enhancement that allows the player to choose the time of day from a menu (or from the clock on the HUD itself) would be super cool. I'm guessing this would have to allow for either the text based clock or for the graphic clock to work though... (but maybe letting the user temporarily modify the time of day in this menu and then restore it if they cancel out would accomplish the same effect).

I figure the branch at line 5815 of z_message.c could contain some of this behavior alongside any code introduced in the enhancements section of the code. 
I'd be fine with trying to implement this myself if ya'll are busy on randomizer content, other enhancements, and stuff like that! I've been looking through the code and I think I have a few ideas of where I would start.

I mainly wanted to put this request out there in case someone was already working on it so no one else (like me) tried to implement it and got in the way haha.

(Feel free to resolve this if this was the wrong place to put this)

**Comment by garrettjoecox** [comment]
This is for sure planned, I think I am the only one who has begun looking into it, but I'm also swamped with other things so don't wait for me if you want to take a crack at it yourself.

After some investigation, my plan was to re-use the same patterns that the banker and bomber code input use, with text boxes that allow you to change the numbers with the analog stick.

All that said, going to convert this to a discussion, to keep our issues section limited to bugs. If you have any questions/problems feel free to ask for help in the #2s2h-development channel in discord!

**Comment by Archez** [comment]
Given we have (currently) two ways to represent the clock (original + text only), I think a solution that doesn't attempt to "integrate" with the clock directly is probably for the best. At least in the sense of not having to code up two solutions. Especially if we ever add a third clock display to match the 3ds clock (all days displayed at once).

**Reply by Dumbledork01** [reply]
I agree with this. I'm thinking temporarily storing the user's time prior to their modification is the easiest thing. Then, if they cancel out, it can be restored. Also, not allowing them to go beneath their initial time. Then, the only thing being modified is the current time rather than any HUDs. I'm curious to see if the HUDs would automatically update or not in this menu (I'd like to still have them show if possible).
The only issue I see with this is if the game freaks out when the user is selecting a time past 6 pm or 6 am and automatically moves to the next day, but I'm sure the flags can be messed with so if they're in that menu it won't trigger those effects.

**Comment by Dumbledork01** [comment]
Took a further look into this in z_message.c: looks like func_80148D64 & func_80149048 are both for selecting rupee amounts to deposit or withdraw from the bank. func_801491DC has 5 numbers that can be between 1 & 5, so it's probably the bomber code entry. func_80149454 has 3 numbers between 0 & 9, so it's probably the lottery. func_801496C8 has 6 numbers between 0 & 3, I don't know what the heck this is...

msgCtx->unk120C2 seems to represent which index of a changeable number is currently selected in the textbox.
msgCtx->unk12054 seems to be a variable storing the selections for lottery numbers, bomber codes, and whatever other text entries exist in the code (except for banker rupee counts).

I think the Song of Double Time could use a similar format to func_801491DC or func_80149454 (the bomber code or lottery code) since both of them are using similar formats. Custom text will also need to be introduced for this, but for now, I think implementing it with their text would be a good proof-of-concept.
